contracts/cw20-merkle-airdrop/src/contract.rs

```diff
@@ -0,0 +1,486 @@
+
+    let user_input = format!("{}{}", info.sender, amount);
+    let hash = sha3::Keccak256::digest(user_input.as_bytes())
+        .as_slice()
+        .try_into()
+        .map_err(|_| ContractError::WrongLength {})?;
+
+    let hash = proof.into_iter().try_fold(hash, |hash, p| {
+        let mut proof_buf = [0; 32];
+        hex::decode_to_slice(p, &mut proof_buf)?;
+        let mut hashes = [hash, proof_buf];
+        hashes.sort_unstable();
+        sha3::Keccak256::digest(&hashes.concat())
+            .as_slice()
+            .try_into()
+            .map_err(|_| ContractError::WrongLength {})
+    })?;
+
+    let mut root_buf: [u8; 32] = [0; 32];
+    hex::decode_to_slice(merkle_root, &mut root_buf)?;
+    if root_buf != hash {
+        return Err(ContractError::VerificationFailed {});
+    }
+
+    // Update claim index to the current stage
+    CLAIM.save(deps.storage, (&info.sender, stage.into()), &true)?;
+
+    let msgs: Vec<SubMsg> = vec![SubMsg::new(WasmMsg::Execute {
+        contract_addr: config.cw20_token_address.to_string(),
+        funds: vec![],
+        msg: to_binary(&Cw20ExecuteMsg::Transfer {
+            recipient: info.sender.to_string(),
+            amount,
+        })?,
+    })];
+    Ok(Response {
+        messages: msgs,
+        attributes: vec![
+            attr("action", "claim"),
+            attr("stage", stage),
+            attr("address", info.sender),
+            attr("amount", amount),
+        ],
+        data: None,
+        events: vec![],
+    })
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
+    match msg {
+        QueryMsg::Config {} => to_binary(&query_config(deps)?),
+        QueryMsg::MerkleRoot { stage } => to_binary(&query_merkle_root(deps, stage)?),
+        QueryMsg::LatestStage {} => to_binary(&query_latest_stage(deps)?),
+        QueryMsg::IsClaimed { stage, address } => {
+            to_binary(&query_is_claimed(deps, stage, address)?)
+        }
+    }
+}
+
+pub fn query_config(deps: Deps) -> StdResult<ConfigResponse> {
+    let cfg = CONFIG.load(deps.storage)?;
+    Ok(ConfigResponse {
+        owner: cfg.owner.to_string(),
+        cw20_token_address: cfg.cw20_token_address.to_string(),
+    })
+}
+
+pub fn query_merkle_root(deps: Deps, stage: u8) -> StdResult<MerkleRootResponse> {
+    let merkle_root = MERKLE_ROOT.load(deps.storage, U8Key::from(stage))?;
+    let resp = MerkleRootResponse { stage, merkle_root };
+
+    Ok(resp)
+}
+
+pub fn query_latest_stage(deps: Deps) -> StdResult<LatestStageResponse> {
+    let latest_stage = STAGE.load(deps.storage)?;
+    let resp = LatestStageResponse { latest_stage };
+
+    Ok(resp)
+}
+
+pub fn query_is_claimed(deps: Deps, stage: u8, address: String) -> StdResult<IsClaimedResponse> {
+    let key: (&Addr, U8Key) = (&deps.api.addr_validate(&address)?, stage.into());
+    let is_claimed = CLAIM.load(deps.storage, key)?;
```

Comment:
This will return error if it is not claimed, when we want false.

Try `let is_claimed = CLAIM.may_load(deps.storage, key)?.unwrap_or(false);`